File: 2020/day01/main.c
/**
 * day 01 2020
 * https://adventofcode.com/2020/day/1
 *
 * compile:
 *    debug:   gcc main.c -o main -g -fsanitize=address -Wall
 *    release: gcc main.c -o main -O2
 * 
 * run:         ./main < input
 **/
#include <stdio.h>
#include <stdint.h>

#define LEN 200

uint32_t part1(uint16_t *n)
{
    uint8_t i, j;
    for (i = 0; i < LEN; i++) {
        for (j = 0; j < LEN; j++) {
            if (i == j) continue;
            if (n[i] + n[j] == 2020) 
                return n[i]*n[j];
        }
    }
    return 0;
}


uint32_t part2(uint16_t *n)
{
    uint8_t i, j, k;
    for (i = 0; i < LEN; i++) {
        for (j = 0; j < LEN; j++) {
            for (k = 0; k < LEN; k++) {
                if (i == j && i == k) continue;
                if (n[i] + n[j] + n[k] == 2020)
                    return n[i]*n[j]*n[k];
            }
        }
    }
    return 0;
}


int main(void)
{

    uint16_t nums[LEN] = {0};
    uint8_t i = 0;

    while (fscanf(stdin, "%hd", &nums[i++]) == 1);

	uint32_t p1 = part1(nums);
	uint32_t p2 = part2(nums);

    printf("part 1: %u\n", p1);
    printf("part 2: %u\n", p2);

    return 0;
}

File: 2020/day01/main.go
package main

import (
	"bufio"
	"os"
	"fmt"
	"strings"
)

func part1(n []int) int {
	for _, i := range n {
		for _, j := range n {
			if i + j == 2020 {
				return i * j
			}
		}
	}
	return -1
}

func part2(n []int) int {
	for _, i := range n {
		for _, j := range n {
			for _, k := range n {
				if i + j +k == 2020 {
					return i * j * k
				}
			}
		}
	}
	return -1
}

func main() {
	var nums []int
	var n int

	scan := bufio.NewScanner(os.Stdin)
	for scan.Scan() {
		f := strings.NewReader(scan.Text())
		fmt.Fscanf(f, "%d", &n)
		nums = append(nums, n)
	}

	a := part1(nums)
	b := part2(nums)

	fmt.Println("part 1:", a)
	fmt.Println("part 2:", b)
}
